OpenRA.Mods.Common/Traits/Buildings/Gate.cs

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class GateInfo : UpgradableTraitInfo, ITraitInfo, Requires<RenderSimpleInfo>
+	{
+		public readonly string OpenSequence = "open";
+		public readonly string ClosingSequence = "closing";
+		public readonly string ClosedSequence = "closed";
+
+		public readonly string OpeningSound = "";
+		public readonly string ClosingSound = "";
+
+		[Desc("- means it blocks when closed.")]
+		public readonly string Footprint = "-";
+		public readonly CVec Dimensions = new CVec(1, 1);
+
+		public readonly WRange ScanRange = WRange.FromCells(3);
+
+		public object Create(ActorInitializer init) { return new Gate(init, this); }
+	}
+
+	public class Gate : ITick, IOccupySpace, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly GateInfo info;
+		readonly Actor self;
+		readonly RenderSimple renderSimple;
+
+		bool cachedOpen;
+
+		CPos topLeft;
+		WPos center;
+
+		Pair<CPos, SubCell>[] blockedCells;
+
+		public Gate(ActorInitializer init, GateInfo info)
+		{
+			this.info = info;
+			this.self = init.Self;
+
+			renderSimple = self.Trait<RenderSimple>();
+
+			topLeft = init.Get<LocationInit, CPos>();
+			center = init.World.Map.CenterOfCell(topLeft) +
+				((init.World.Map.CenterOfCell(CPos.Zero + new CVec(info.Dimensions.X, info.Dimensions.Y))
+					- init.World.Map.CenterOfCell(new CPos(1, 1))) / 2);
+
+			blockedCells = BlockedCells().Select(c => Pair.New(c, SubCell.FullCell)).ToArray();
+		}
+
+		public CPos TopLeft { get { return topLeft; } }
+		public WPos CenterPosition { get { return center; } }
+		public IEnumerable<Pair<CPos, SubCell>> OccupiedCells() { return cachedOpen ? new Pair<CPos, SubCell>[0] : blockedCells; }
+
+		public IEnumerable<CPos> BlockedCells()
+		{
+			var footprint = info.Footprint.Where(x => !char.IsWhiteSpace(x)).ToArray();
+			foreach (var tile in FootprintUtils.TilesWhere(self.Info.Name, (CVec)info.Dimensions, footprint, a => a == '-'))
+				yield return tile + topLeft;
+		}
+
+		public void AddedToWorld(Actor self)
+		{
+			self.World.ActorMap.AddInfluence(self, this);
+			self.World.ActorMap.AddPosition(self, this);
+			self.World.ScreenMap.Add(self);
+		}
+
+		public void RemovedFromWorld(Actor self)
+		{
+			self.World.ActorMap.RemoveInfluence(self, this);
+			self.World.ActorMap.RemovePosition(self, this);
+			self.World.ScreenMap.Remove(self);
+		}
+
+		public void Tick(Actor self)
+		{
+			if (self.IsDisabled())
+				return;
+
+			var open = UnitsInRange().Any(a => !a.Owner.NonCombatant && a.AppearsFriendlyTo(self) && a.HasTrait<Mobile>());
```

Comment:
I changed behavior to keep this open until enemies arrive to hide the current limitation a bit.